scalyr_agent/sharded_copying_manager/copying_manager.py

```diff
@@ -198,30 +204,48 @@ def generate_status(self, warn_on_rate_limit=False):
                     result.last_attempt_time, status.last_attempt_time
                 )
 
+            # the next fields are a sum of the same fields from all workers.
             if status.last_attempt_size:
                 result.last_attempt_requests_overall_size += status.last_attempt_size
 
-            # Get the overall status about last responses from all workers.
-            # If at least one response is not successful then the result is not successful too.
-            if status.last_response_status is not None:
-                # do not overwrite the result if we already have a bad response.
-                if result.all_responses_successful is not False:
-                    result.all_responses_successful = (
-                        status.last_response_status == "success"
-                    )
-
-            # Get the overall status about the health checks from all workers.
-            # If at least one check is not successful then the result is not successful too.
-            if status.health_check_result is not None:
-                # do not overwrite the result if there is a bad health check.
-                if result.all_health_checks_good is not False:
-                    result.all_health_checks_good = status.health_check_result == "Good"
-
-            # the next fields are a sum of the same fields from all workers.
             result.total_bytes_uploaded += status.total_bytes_uploaded
 
             result.total_errors += status.total_errors
 
+        # get overall information about last responses of all workers in the worker pool.
+        # making this as a set should help in further decision making process.
+        responses = {
+            # compare each response status for "success" and store the boolean result.
+            # NOTE: we do not compare None values, they are stored as are.
+            None
+            if s.last_response_status is None
+            else s.last_response_status == "success"
+            for s in worker_statuses
+        }
+
+        if False in responses:
```

Comment:
I see at least four places where you have this pattern of accumulating the responses together, handling the failure case and None cases in this manner.  Maybe we should just have a helper method that does this.  Something like:

```
def accumulate_responses(responses, result_on_at_least_one_fail=False, result_on_all_succeed=True):
    if False in response:
        return result_on_at_least_one_fail
    elif None in responses:
        return None
    else:
         return result_on_all_succeeded
```

I think that would work here and below but give it a try.

Reply:
Thanks, that helped to reduce the amount of code.